### ISSUE TITLE
Typing errors in `Tesla.Client`

### ISSUE THREAD

**Issue by halostatue** [teamon/tesla#447]
I was trying to use `hammox` (a typespec-aware wrapper around `mox`) with Tesla, but the types differ from the values.

Here’s the type in `Tesla.Client`:

```elixir
  @type adapter :: module | {module, any} | (Tesla.Env.t() -> Tesla.Env.result())
  @type middleware :: module | {module, any}
  # This is from Tesla.Env
  @type stack :: [{atom, atom, any} | {atom, atom} | {:fn, (t -> t)} | {:fn, (t, stack -> t)}]

  @type t :: %__MODULE__{
          pre: Tesla.Env.stack(),
          post: Tesla.Env.stack(),
          adapter: adapter | nil
        }
```

Here’s what’s returned in an `inspect`:

```elixir
%Tesla.Client{
  adapter: {Tesla.Adapter.Hackney, :call,
   [[ssl_options: [versions: [:"tlsv1.2"]]]]},
  fun: nil,
  post: [],
  pre: [
    {Tesla.Middleware.Timeout, :call, [[timeout: 30000]]},
    {Tesla.Middleware.Compression, :call, [[]]},
    {Tesla.Middleware.JSON, :call, [[]]},
    {Tesla.Middleware.Logger, :call, [[]]}
  ]
}
```

It feels like `adapter` should be specified as:

```elixir
@type adapter :: {module, :call, any} | (Tesla.Env.t() -> Tesla.Env.result())
```

**Comment by gaiabeatrice** [comment]
Hi @halostatue,
I am not the creator of the library. I am just an enthusiastic user. I also successfully use Tesla and I use Hammox for testing its calls.

Here is an example of one of my calls and it successfully works, I thought it can help you.

```elixir
    TeslaMock
    |> expect(:call, fn
      %Tesla.Env{
        body: "{\"foo\":42}",
        headers: [{"content-type", "application/json"}],
        url: ^url_success
      },
      [] ->
        TeslaHelper.response(
          body: ~s|{"success": true}|,
          method: :post,
          url: url_success,
          status: 200,
          headers: [{"content-type", "application/json"}]
        )
    end)
```

And this is my Helper module looks like:
```elixir
defmodule MyApp.Support.TeslaHelper do
  def response(args) do
    {:ok,
     %Tesla.Env{
       __client__: Keyword.get(args, :tesla_client, %Tesla.Client{}),
       status: Keyword.get(args, :status),
       body: Keyword.get(args, :body),
       method: Keyword.get(args, :method, :get),
       url: Keyword.get(args, :url, ""),
       headers: Keyword.get(args, :headers, [])
     }}
  end
end

```

**Comment by halostatue** [comment]
My guess is that this is working because you’re returning a blank `Tesla.Client` as part of the response, whereas I’m returning the values that are passed _in_ for the `__client__` value. I’m doing something like

```elixir
expect(TeslaMock, :call, fn env, _opts ->
    %{env | body: ~s({"success": true}), status: 200}
  end)
```

Specifically, Hammox is validating that `Tesla.Client.adapter` conforms to `module | {module, any} | (Tesla.Env.t() -> Tesla.Env.result())`, which it doesn’t. It conforms to `mfa` (`{module, :call, any}` to be precise). The typespec of `stack` is more correct than the typespec of `adapter` or `middleware`.

**Comment by chulkilee** [comment]
That's because `adapter` type is for argument being passed to Tesla functions (for an adapter), but Tesla keeps it with runtime - see following lines

https://github.com/teamon/tesla/blob/v1.4.0/lib/tesla/builder.ex#L174

https://github.com/teamon/tesla/blob/v1.4.0/lib/tesla/builder.ex#L197

So the fix would be one of followings

- Fix the `adapter` type in the struct by adding a case for `{module, :call, [opts]}` - (`adapter: adapter | nil | {module, :call, any}` for instance)
- Change the internal to keep the original `{module, [opts]}` in `adapater` - and when needed convert it to `{module, :call, [opts]}`

**Comment by halostatue** [comment]
If the `mfa` form is an internal implementation detail, then the _right_ choice would probably be to change the internals to keep it as `{module, [opts]}`. The _quick_ choice would be to add `{module, :call, any}` to the specs.

**Comment by chulkilee** [comment]
To make the change backward compatible, we should fix the type spec (changing from `adapter | nil` to `Tesla.Env.stack() | nil`), not changing the value in the struct (although it *is* different from the spec). @halostatue could you test the change would remove the 

Maybe there are slight performance penalty to convert the ma from to mfa if it's being called for every request?

Anyway, created #449 for fixing typespec

**Comment by halostatue** [comment]
Um. `Tesla.Env.stack` is a list (`@type stack :: [{atom, atom, any} | {atom, atom} | {:fn, (t -> t)} | {:fn, (t, stack -> t)}]`). 

I think this would be better as

```elixir
  @type adapter :: {module, :call, any} | module | {module, any} | (Tesla.Env.t() -> Tesla.Env.result())
```

I’ll be able to test this tomorrow.

**Comment by teamon** [comment]
Thank you all ❤️ 💚 💙 💛 

I just merged #449. Please try again with tesla `master`.
